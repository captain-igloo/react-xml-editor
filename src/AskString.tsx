import * as React from 'react';

import { Actions, Xml } from './types';
import { updateNode } from './Util';

export enum AskStringType {
    LONG,
    SHORT,
}

interface Props {
    actions: Actions;
    defaultValue: string;
    id: string[];
    type: AskStringType;
    xml: Xml;
}

interface State {
    value: string;
}

export default class AskString extends React.Component<Props, State> {
    public constructor(props: Props) {
        super(props);
        this.onChange = this.onChange.bind(this);
        this.onSubmit = this.onSubmit.bind(this);
        this.state = {
            value: props.defaultValue,
        };
    }

    public componentDidUpdate(prevProps: Props) {
        const { defaultValue } = this.props;
        if (prevProps.defaultValue !== defaultValue) {
            this.setState({
                value: this.props.defaultValue,
            });
        }
    }

    public render(): React.ReactNode {
        const { type } = this.props;
        return (
            <form onSubmit={ this.onSubmit }>
                { type === AskStringType.LONG ? this.getLongString() : this.getShortString() }
            </form>
        );
    }



    private getShortString(): React.ReactNode {
        const { value } = this.state;

        return (
            <>
                <input
                    aria-label="Value"
                    name="val"
                    className="textbox focusme"
                    onChange={ this.onChange }
                    value={ value }
                />
                <input type="submit" value="OK"/>
            </>
        );
    }

    private getLongString(): React.ReactNode {
        const { value } = this.state;
        return (
            <>
                <textarea
                    name="val"
                    className="textbox focusme"
                    onChange={ this.onChange }
                    value={value}
                />
                <div className="submitline">
                    <input type="submit" value="OK" />
                </div>
            </>
        );
    }

    private onSubmit(e: React.FormEvent<HTMLFormElement>) {
        const { actions, id, xml } = this.props;
        const { value } = this.state;

        actions.setXml(updateNode(xml, id, value));
        actions.showBubble({
            show: false,
        });
        e.preventDefault();
    }

    private onChange(e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) {
        this.setState({
            value: e.target.value,
        });
    }
}
