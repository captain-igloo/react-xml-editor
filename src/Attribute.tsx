import * as React from 'react';

import { BubbleType, Actions, BubbleOptions } from './types';

interface Props {
    actions: Actions;
    element: string;
    id: string[];
    name: string;
    value: string;
}

export default class Attribute extends React.Component<Props> {
    private ref: React.RefObject<HTMLSpanElement | null>;

    public constructor(props: Props) {
        super(props);
        this.ref = React.createRef();
        this.onClickName = this.onClickName.bind(this);
        this.onClickValue = this.onClickValue.bind(this);
    }

    public render(): React.ReactNode {
        const { name, value } = this.props;

        return (
            <span className="attribute">
                <span className="punc"> </span>
                <span className="name" onClick={ this.onClickName }>{ name }</span>
                <span className="punc">=</span>
                <span className="valueContainer" onClick={ this.onClickValue } ref={ this.ref }>
                    <span className="punc">&quot;</span>
                    <span className="value">{ value }</span>
                    <span className="punc">&quot;</span>
                </span>
            </span>
        );
    }

    private onClickName(event: React.MouseEvent): void {
        const { actions, element, id, name, value } = this.props;
        const bubbleOptions: Partial<BubbleOptions> = {
            attribute: name,
            element,
            id,
            show: true,
            type: BubbleType.MENU,
            value,
        };
        if (this.ref.current) {
            const rect = this.ref.current.getBoundingClientRect();
            bubbleOptions.left = rect.left;
            bubbleOptions.top = rect.top;
        }
        actions.showBubble(bubbleOptions);
        event.stopPropagation();
    }

    private onClickValue(event: React.MouseEvent): void {
        const { actions, element, id, name, value } = this.props;
        const bubbleOptions: Partial<BubbleOptions> = {
            attribute: name,
            element,
            id,
            show: true,
            type: BubbleType.ASKER,
            value,
        };
        if (this.ref.current) {
            const rect = this.ref.current.getBoundingClientRect();
            bubbleOptions.left = rect.left;
            bubbleOptions.top = rect.top;
        }
        actions.showBubble(bubbleOptions);
        event.stopPropagation();
    }
}
