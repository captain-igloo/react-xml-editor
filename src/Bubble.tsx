import * as React from 'react';

import { BubbleType, Actions, DocSpec, MenuItemSpec, Xml } from './types';
import { askLongString } from './Util';

interface Props {
    actions: Actions;
    attribute: string;
    docSpec: DocSpec;
    element: string;
    id: string[];
    left: number;
    mode: 'laic' | 'nerd';
    show: boolean;
    top: number;
    type: BubbleType;
    value: string;
    xml: Xml;
}

const formatCaption = (caption: string) => {
    let key = 1;
    return caption.split(/((?:<\/?[^>]+\/?>)|(?:@[^ =]+(?:="[^"]*")?))/).map((piece) => {
        const elementMatches = piece.match(/<(\/?)([^>/]+)(\/?)>/);
        const attrMatches = piece.match(/@(?:([^ ="]+))?=?(?:"([^"]*)")?/);
        if (elementMatches) {
            return (
                <span className="techno" key={ key++ }>
                    <span className="punc">&lt;{ elementMatches[1] }</span>
                    <span className="elName">{ elementMatches[2] }</span>
                    <span className="punc">{ elementMatches[3] }&gt;</span>
                </span>
            );
        } else if (attrMatches) {
            let name;
            if (attrMatches[1]) {
                name = <span className="atName">{ attrMatches[1] }</span>;
            }
            let value;
            if (typeof attrMatches[2] !== 'undefined') {
                let equals;
                if (name) {
                    equals = <span className="punc equals">=</span>;
                }
                value = (
                    <React.Fragment>
                        { equals }
                        <span className="punc">&quot;</span>
                        { attrMatches[2] && <span className="atValue">{ attrMatches[2] }</span> }
                        <span className="punc">&quot;</span>
                    </React.Fragment>
                );
            }

            return (
                <span className="techno" key={ key++ }>
                    { name }
                    { value }
                </span>
            );
        }
        return piece;
    });
};

export default class Bubble extends React.Component<Props> {
    public constructor(props: Props) {
        super(props);
        this.showMenuItem = this.showMenuItem.bind(this);
    }

    public render(): React.ReactNode {
        const {
            id,
            show,
        } = this.props;
        if (!show) {
            return null;
        }
        if (id.length > 2 && id[id.length - 2] === '$') {
            return this.getAttributeBubble();
        } else if (id.length > 2 && id[id.length - 2] === '$$') {
            return this.getElementBubble();
        } else if (id.length > 1 && id[id.length - 1] === '_') {
            return this.getTextBubble();
        }
        return null;
    }

    private getTextBubble(): React.ReactNode {
        const { actions, docSpec, element, id, left, mode, top, value, xml } = this.props;
        const asker = docSpec.elements?.[element]?.asker || askLongString;
        return (
            <div
                className={`xonomyBubble ${mode}`}
                style={{ left, top, display: 'block' }}
            >
                <div className="inside">
                    <div className="xonomyBubbleContent">
                        { asker({
                            actions,
                            defaultValue: value,
                            id,
                            xml,
                        }) }
                    </div>
                </div>
            </div>
        );
    }

    private getElementBubble(): React.ReactNode {
        const { actions, docSpec, element, id, left, mode, top, xml } = this.props;

        const menu = docSpec.elements?.[element]?.menu;
        if (menu) {
            const menuItems = menu.filter(this.showMenuItem).map((menuItemSpec, index) => {
                let icon;
                if (menuItemSpec.icon) {
                    icon = (
                        <span className="icon">
                            <img src={ menuItemSpec.icon } />
                        </span>
                    );
                }
                return (
                    <div
                        className="menuItem focusme"
                        key={ index }
                        onClick={ async () => {
                            actions.setXml(await menuItemSpec.action(xml, id));
                            actions.showBubble({ show: false });
                        } }
                    >
                        { icon }
                        { formatCaption(menuItemSpec.caption) }
                    </div>
                );
            });

            return (
                <div
                    className={`xonomyBubble ${mode}`}
                    style={{ left, top, display: 'block' }}
                >
                    <div className="inside">
                        <div className="xonomyBubbleContent">
                            <div className="menu">
                                {menuItems}
                            </div>
                        </div>
                    </div>
                </div>
            );
        }
        return null;
    }

    private getAttributeBubble(): React.ReactNode {
        const { type } = this.props;
        if (type === BubbleType.ASKER) {
            return this.getAttributeAskerBubble();
        }
        return this.getAttributeMenuBubble();
    }

    private getAttributeMenuBubble(): React.ReactNode {
        const { actions, attribute, docSpec, element, id, left, mode, top, xml } = this.props;
        const menu = docSpec.elements?.[element]?.attributes?.[attribute]?.menu;
        if (menu) {
            const menuItems = menu.filter(this.showMenuItem).map((menuItemSpec, index) => (
                <div
                    className="menuItem focusme"
                    key={ index }
                    onClick={ async () => {
                        actions.setXml(await menuItemSpec.action(xml, id));
                        actions.showBubble({ show: false });
                    } }
                >
                    { menuItemSpec.caption }
                </div>
            ));

            return (
                <div className={`xonomyBubble ${mode}`} style={{ left, top, display: 'block' }}>
                    <div className="inside">
                        <div className="xonomyBubbleContent">
                            <div className="menu">
                                {menuItems}
                            </div>
                        </div>
                    </div>
                </div>
            );
        }
        return null;
    }

    private getAttributeAskerBubble(): React.ReactNode {
        const { actions, attribute, docSpec, element, id, left, mode, top, value, xml } = this.props;
        const attributeSpec = docSpec.elements?.[element]?.attributes?.[attribute];
        if (attributeSpec && attributeSpec.asker) {
            const asker = attributeSpec.asker;
            return (
                <div className={`xonomyBubble ${mode}`} style={{ left, top, display: 'block' }}>
                    <div className="inside">
                        <div className="xonomyBubbleContent">
                            { asker({
                                actions,
                                defaultValue: value,
                                id,
                                xml,
                            }) }
                        </div>
                    </div>
                </div>
            );
        }
        return null;
    }

    private showMenuItem(menuItemSpec: MenuItemSpec): boolean {
        const { id, xml } = this.props;
        if (menuItemSpec.hideIf) {
            return !menuItemSpec.hideIf(xml, id);
        }
        return true;
    }
}
