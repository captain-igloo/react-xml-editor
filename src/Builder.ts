import * as builder from 'xmlbuilder';

import { Xml } from './types';

export interface BuilderOptions {
    doctype?: {
        pubID?: string;
        sysID?: string;
    };
    headless?: boolean;
    xmldec?: {
        encoding?: string;
        standalone?: boolean;
        version?: '1.0';
    };
}

export default class Builder {
    private attrKey = '$';

    private charKey = '_';

    private childKey = '$$';

    private textKey = '__text__';

    private headless: boolean;

    private xmldec: {
        encoding?: string;
        standalone?: boolean;
        version?: string;
    };

    private doctype?: {
        pubID?: string;
        sysID?: string;
    };

    public constructor(options: BuilderOptions) {
        this.headless = options.headless || false;
        this.xmldec = options.xmldec || {
            encoding: 'UTF-8',
            standalone: true,
            version: '1.0',
        };
        this.doctype = options.doctype;
    }

    public buildObject(rootObj: Xml): string {
        let rootName;

        if (Object.keys(rootObj).length === 1) {
            rootName = Object.keys(rootObj)[0];
        } else {
            throw new Error('invalid');
        }

        const rootElement = builder.create(rootName, this.xmldec, this.doctype, {
            headless: this.headless,
        });

        return this.render(rootElement, rootObj[rootName]).end({
            indent: '  ',
            newline: '\n',
            pretty: true,
        });
    }

    private render(element: builder.XMLElement, obj: any) {
        console.log('OBJ', obj);
        if (Array.isArray(obj)) {
            obj.forEach((child) => {
                if ('#name' in child) {
                    if (child['#name'] === this.textKey) {
                        element = element.txt(child[this.charKey]);
                    } else {
                        element = this.render(element.ele(child['#name']), child).up();
                    }
                }
            });
        } else {
            if (this.attrKey in obj) {
                const child = obj[this.attrKey];
                Object.keys(child).forEach((attr) => {
                    const value = child[attr];
                    element = element.att(attr, value);
                });
            }
            if (this.childKey in obj) {
                const child = obj[this.childKey];
                if (Array.isArray(child)) {
                    element = this.render(element, child);
                }
            }
        }
        return element;
    }
}
