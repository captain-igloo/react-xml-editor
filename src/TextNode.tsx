import * as React from 'react';

import { Actions, BubbleOptions } from './types';

interface Props {
    actions: Actions;
    element: string;
    id: string[];
    text: string;
}

export default class TextNode extends React.Component<Props> {
    private ref: React.RefObject<HTMLDivElement | null>;

    public constructor(props: Props) {
        super(props);
        this.onClick = this.onClick.bind(this);
        this.ref = React.createRef();
    }

    public render(): React.ReactNode {
        const { text } = this.props;
        return (
            <div
                className="textnode focusable"
                onClick={this.onClick}
                ref={this.ref}
            >
                <span className="value">{ text }</span>
            </div>
        );
    }

    private onClick(event: React.MouseEvent): void {
        const { actions, element, id, text } = this.props;
        const bubbleOptions: Partial<BubbleOptions> = {
            id,
            element,
            show: true,
            value: text,
        };
        if (this.ref.current) {
            const rect = this.ref.current.getBoundingClientRect();
            bubbleOptions.left = rect.left;
            bubbleOptions.top = rect.top;
        }
        actions.showBubble(bubbleOptions);
        event.stopPropagation();
    }
}
