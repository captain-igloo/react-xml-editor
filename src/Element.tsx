import * as React from 'react';

import Attributes from './Attributes';
import Collapsoid from './Collapsoid';
import TextNode from './TextNode';
import { Actions, BubbleOptions, Element as ElementConfig, Xml } from './types';
import { push, updateNode } from './Util';

interface Props {
    actions: Actions;
    attributes?: { [key: string]: string };
    childElements?: ElementConfig[];
    collapsed?: boolean;
    id: string[];
    name: string;
    xml: Xml;
}

export default class Element extends React.Component<Props> {
    private ref: React.RefObject<HTMLDivElement | null>;

    public constructor(props: Props) {
        super(props);
        this.onCollapse = this.onCollapse.bind(this);
        this.onClick = this.onClick.bind(this);
        this.ref = React.createRef();
    }

    public render(): React.ReactNode {
        const {
            actions,
            childElements,
            collapsed,
            id,
            name,
            xml,
        } = this.props;

        const elements: React.ReactNode[] = [];
        if (childElements) {
            childElements.forEach((childElement: ElementConfig, index: number) => {
                if (childElement['#name'] !== '__text__') {
                    elements.push(
                        <Element
                            actions={actions}
                            attributes={childElement.$}
                            childElements={childElement.$$}
                            collapsed={childElement['#collapsed']}
                            id={push(id, '$$', `${index}`)}
                            key={index}
                            name={childElement['#name']}
                            xml={xml}
                        />,
                    );
                } else if (childElement._) {
                    elements.push(
                        <TextNode
                            actions={actions}
                            element={name}
                            id={push(id, '$$', `${index}`, '_')}
                            key={index}
                            text={childElement._}
                        />,
                    );
                }
            });
        }

        const childrenContainer = (
            <React.Fragment>
                <span className="prominentChildren"/>
                <span className="childrenCollapsed focusable">
                    {childElements && <Collapsoid elements={ childElements }/> }
                </span>
                <div className="children">
                    { elements }
                </div>
            </React.Fragment>
        );

        let elementClass = '';
        if (elements.length === 0) {
            elementClass = ' noChildren';
        }
        let expandCollapseLabel = 'Collapse';
        if (collapsed) {
            elementClass += ' collapsed';
            expandCollapseLabel = 'Expand';
        }

        const openingTag = (
            <span className="tag opening">
                <span className="punc">&lt;</span>
                <span className="name" onClick={ this.onClick } ref={ this.ref }>{ name }</span>
                { this.getAttributes() }
                <span className="punc slash">/</span>
                <span className="punc">&gt;</span>
            </span>
        );

        let closingTag;
        if (elements.length > 0) {
            closingTag = (
                <span className="tag closing">
                    <span className="punc">&lt;</span>
                    <span className="punc slash">/</span>
                    <span className="name">{ name }</span>
                    <span className="punc">&gt;</span>
                </span>
            );
        }

        return (
            <React.Fragment>
                <div className={`element${elementClass}`}>
                    <span className="connector">
                        <span className="plusminus" onClick={ this.onCollapse } title={ expandCollapseLabel } />
                        <span className="draghandle" draggable="true" />
                    </span>
                    { openingTag }
                    { childrenContainer }
                    { closingTag }
                </div>
            </React.Fragment>
        );
    }

    private onClick(event: React.MouseEvent) {
        const { actions, id, name } = this.props;
        const bubbleOptions: Partial<BubbleOptions> = {
            id,
            element: name,
            show: true,
            value: name,
        };
        if (this.ref.current) {
            const rect = this.ref.current.getBoundingClientRect();
            bubbleOptions.left = rect.left;
            bubbleOptions.top = rect.top;
        }
        actions.showBubble(bubbleOptions);
        event.stopPropagation();
    }

    private onCollapse() {
        const { actions, collapsed, id, xml } = this.props;
        const collapseId = push(id, '#collapsed');
        actions.showBubble({
            show: false,
        });
        actions.setXml(updateNode(xml, collapseId, !collapsed));
    }

    private getAttributes() {
        const { actions, attributes, id, name } = this.props;
        if (attributes) {
            return (
                <Attributes
                    actions={ actions }
                    attributes={ attributes }
                    element={ name }
                    id={ id }
                />
            );
        }
        return null;
    }
}
