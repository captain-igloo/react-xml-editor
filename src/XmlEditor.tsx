import * as React from 'react';

import Bubble from './Bubble';
import Element from './Element';
import Parser from './Parser';
import {
    BubbleType,
    Actions,
    BubbleOptions,
    DocSpec,
    Xml,
} from './types';

type DefaultProps = {
    mode: 'laic' | 'nerd';
};

type Props = {
    docSpec: DocSpec;
    onChange?: () => void;
    ref: React.RefObject<XmlEditor | null>;
    xml: string;
} & Partial<DefaultProps>;

interface State {
    bubble: BubbleOptions;
    xml?: Xml;
}

export default class XmlEditor extends React.Component<Props & DefaultProps, State> {
    static defaultProps: DefaultProps = {
        mode: 'nerd',
    };

    public constructor(props: Props & DefaultProps) {
        super(props);
        this.setXml = this.setXml.bind(this);
        this.showBubble = this.showBubble.bind(this);
        this.onClick = this.onClick.bind(this);
        this.state = {
            bubble: {
                attribute: '',
                element: '',
                id: [],
                left: 0,
                show: false,
                top: 0,
                type: BubbleType.ASKER,
                value: '',
            },
        };
    }

    public componentDidMount() {
        const { xml } = this.props;
        const parser = new Parser();
        parser.parseString(xml).then((result) => {
            this.setState({
                xml: result,
            });
        });
    }

    public getXml() {
        const { xml } = this.state;
        return xml;
    }

    private onClick() {
        this.showBubble({
            show: false,
        });
    }

    private setXml(xml: Xml) {
        const { onChange } = this.props;

        this.setState({
            xml,
        }, () => {
            if (onChange) {
                onChange();
            }
        });
    }

    private showBubble(askOptions: Partial<BubbleOptions>): void {
        this.setState((prevState: State) => ({
            bubble: Object.assign(prevState.bubble, askOptions),
        }));
    }

    public render(): React.ReactNode {
        const { mode } = this.props;
        return (
            <>
                <div className={`xonomy ${mode}`} onClick={this.onClick}>
                    { this.getRootNode() }
                </div>
                { this.getBubble() }
            </>
        );
    }

    private getActions(): Actions {
        return {
            setXml: this.setXml,
            showBubble: this.showBubble,
        };
    }

    private getBubble(): React.ReactNode {
        const { docSpec, mode } = this.props;
        const { bubble, xml } = this.state;
        if (xml) {
            return (
                <Bubble
                    actions={this.getActions()}
                    attribute={bubble.attribute}
                    docSpec={docSpec}
                    element={bubble.element}
                    id={bubble.id}
                    left={bubble.left}
                    mode={mode}
                    show={bubble.show}
                    top={bubble.top}
                    type={bubble.type}
                    value={bubble.value}
                    xml={xml}
                />
            );
        }
        return null;
    }

    private getRootNode(): React.ReactNode {
        const { xml } = this.state;
        if (xml) {
            const key = Object.keys(xml)[0];
            return (
                <Element
                    actions={this.getActions()}
                    attributes={xml[key].$}
                    childElements={xml[key].$$}
                    collapsed={xml[key]['#collapsed']}
                    id={[key]}
                    name={key}
                    xml={xml}
                />
            );
        }
        return null;
    }
}
