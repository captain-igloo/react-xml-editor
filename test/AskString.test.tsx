import { describe, expect, jest, test } from '@jest/globals';
import { render, fireEvent } from '@testing-library/react';
import * as React from 'react';

import AskString, { AskStringType } from '../src/AskString';
import { Actions, Xml } from '../src/types';

describe('AskString component', () => {
    test('ask short string renders properly', () => {
        const { container } = render(
            <AskString
                actions={{} as Actions}
                defaultValue="defaultValue"
                id={['id']}
                type={ AskStringType.SHORT }
                xml={ {} as Xml }
            />,
        );
        expect(container).toMatchSnapshot();
    });

    test('Click OK should set xml and hide bubble', () => {
        const setXml = jest.fn();
        const showBubble = jest.fn();
        const { getByDisplayValue, getByLabelText } = render(
            <AskString
                actions={{ setXml, showBubble } as any}
                defaultValue="defaultValue"
                id={['id']}
                type={ AskStringType.SHORT }
                xml={ {} as Xml }
            />,
        );
        fireEvent.change(getByLabelText('Value'), {
            target: {
                value: 'new value',
            },
        });
        fireEvent.click(getByDisplayValue('OK'));
        expect(setXml).toHaveBeenCalledWith({
            id: 'new value',
        });
        expect(showBubble).toHaveBeenCalledWith({
            show: false,
        });
    });

    test('asdf', () => {
        const { container, rerender } = render(
            <AskString
                actions={{ } as any}
                defaultValue="defaultValue"
                id={['id']}
                type={ AskStringType.LONG }
                xml={ {} as Xml }
            />,
        );
        rerender(
            <AskString
                actions={{ } as any}
                defaultValue="newValue"
                id={['id']}
                type={ AskStringType.LONG }
                xml={ {} as Xml }
            />,
        );
        expect(container).toMatchSnapshot();
        expect(true).toBe(true);
    });
});
