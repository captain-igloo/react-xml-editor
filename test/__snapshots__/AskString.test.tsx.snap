// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`AskString component asdf 1`] = `
<div>
  <form>
    <textarea
      class="textbox focusme"
      name="val"
    >
      newValue
    </textarea>
    <div
      class="submitline"
    >
      <input
        type="submit"
        value="OK"
      />
    </div>
  </form>
</div>
`;

exports[`AskString component ask short string renders properly 1`] = `
<div>
  <form>
    <input
      aria-label="Value"
      class="textbox focusme"
      name="val"
      value="defaultValue"
    />
    <input
      type="submit"
      value="OK"
    />
  </form>
</div>
`;
